interpreter/cling/lib/Interpreter/LookupHelper.cpp

```diff
@@ -81,17 +86,48 @@ namespace cling {
     }
     assert(!code.empty()&&"prepareForParsing should only be called when needd");
 
-    //
-    //  Create a fake file to parse the type name.
-    //
-    std::unique_ptr<llvm::MemoryBuffer>
-      SB = llvm::MemoryBuffer::getMemBufferCopy(code.str() + "\n",
-                                                bufferName.str());
-    SourceLocation NewLoc = Interp->getNextAvailableLoc();
-    FileID FID = S.getSourceManager().createFileID(std::move(SB),
-                                                   SrcMgr::C_User,
-                                                   /*LoadedID*/0,
-                                                   /*LoadedOffset*/0, NewLoc);
+    // Create a fake file to parse the type name.
+    FileID FID;
+    llvm::hash_code hashedCode = llvm::hash_value(code);
+    auto cacheItr = m_LH.m_ParseBufferCache.find(hashedCode);
+    SourceLocation NewLoc;
+    SourceManager& SM = S.getSourceManager();
+    bool CacheIsValid = false;
+    if (cacheItr != m_LH.m_ParseBufferCache.end()) {
+      SourceLocation FileStartLoc =
+        SourceLocation::getFromRawEncoding(cacheItr->second);
+      FID = SM.getFileID(FileStartLoc);
+
+      bool Invalid = true;
+      llvm::StringRef FIDContents = SM.getBuffer(FID, &Invalid)->getBuffer();
+      assert(!Invalid);
+
+      // Compare the contents of the cached buffer and the string we should
+      // process. If there are hash collisions this assert should trigger
+      // making it easier to debug.
+      CacheIsValid = FIDContents.equals(llvm::StringRef(code.str() + "\n"));
+      assert(CacheIsValid && "Hash collision!");
```

Comment:
Shouldn't this be Error? Hash will just collide when it's release

Reply:
Consider adding the following comment here or a few lines above.
```
Discussion on this code implicit handling of unloading.

The relationship is as follows: FileID is a source manager concept wrapping a FileEntry and ContentCache.
The unloading records the list of FileIDs it needs to unload. Then it access their ContentCache and sets it to 0 (still have not figured out how to check that).
The FileID is valid and if the hash lookup finds it, it will rejected because the comparison of its contents (which is now just 0) with the 'code' will indicate a difference.

At the moment we have not yet produced an example that triggers the 2 asserts (on ```Invalid``` and on ```CacheIsValid```), for example the following:

```
cat T.C 
struct S {};

root -l -b
root [0] #undef TRY
root [1] .I /home/vagrant/workspace/root/src/interpreter/cling/include/
root [2] .I /home/vagrant/workspace/root/src/interpreter/llvm/src/include/
root [3] .I /home/vagrant/workspace/root/src/interpreter/llvm/src/tools/clang/include
root [4] .I /usr/lib/llvm-5.0/include
root [5] #include "cling/Interpreter/Interpreter.h"
root [6] #include "cling/Interpreter/LookupHelper.h"
root [7] cling::runtime::gCling->getLookupHelper().findType("S", cling::LookupHelper::NoDiagnostics).isNull()
(bool) true
root [8] #include "T.C"
root [9] cling::runtime::gCling->getLookupHelper().findType("S", cling::LookupHelper::NoDiagnostics).isNull()
(bool) false
root [10] .undo 2
root [11] cling::runtime::gCling->getLookupHelper().findType("S", cling::LookupHelper::NoDiagnostics).isNull()
(bool) true
root [12] .q
```
did not trigger the asserts.

As a side, for the case of real files, loading, unloading and reloading the same (modified) will result in 2 distinct FID. 
```